Optimizes Grid designMode

jbrundage opened the pull request:
Design Mode dropCells are now drawn with a canvas element. This allows it to load much faster and support any realistic number of columns/rows without lag.

Three new publish params:
extraDesignModeWidth - offers control of the number of additional columns added in designMode
extraDesignModeHeight - offers control of the number of additional rows added in designMode
cellDensity - offers control of the number of cells per standard cell

Give it a try:
https://rawgit.com/jbrundage/Visualization/Grid_Optimizing_designMode/demos/dermatology.html?src/layout/Grid&designMode=true

@GordonSmith @mlzummo @dtsnell4 @drealeed Please Review

Signed-off-by: Jay Brundage <jaman.brundage@lexisnexis.com>

GordonSmith commented:
Nice. Two comments:
1.  The grids don't turn off when not in design mode.
2.  This is a request - in designMode can you re-use the selection bag (which should now be in Util) to allow selection of the widgets?  Then the propertyEditor should only show the properties for the selected widgets - and if you select more than 1 you can enable your merged property mode in the property editor????

jbrundage commented:
@GordonSmith The the canvas grid is now removed when exiting designMode.

As for the SelectionBag, I'll create an issue for that, but can these enhancements be merged to master?

GordonSmith commented on src/layout/Grid.js:
```diff
@@ -65,23 +69,24 @@
         rowSpan = rowSpan || 1;
         colSpan = colSpan || 1;
         title = title || "";
+        var mult = this.cellDensity() ? this.cellDensity() : 1;
```
Is this needed?

jbrundage replied:
The condition on that line is probably not needed. It's just intended to allow cellDensity = null or cellDensity = 0

GordonSmith replied:
We don't seem to worry about that for any of the other published params.

GordonSmith commented on src/layout/Grid.js:
```diff
@@ -112,6 +117,21 @@
         return retVal;
     };
     
+    Grid.prototype.updateCellMultiples = function () {
```
Can this get called more than once?

jbrundage replied:
It is called on every .update()

GordonSmith commented on src/layout/Grid.js:
```diff
@@ -112,6 +117,21 @@
     
+    Grid.prototype.updateCellMultiples = function () {
+        var context = this;
+        this.content().forEach(function (cell) {
```
I suspect this whole loop can be avoided most of the time:
```
if (this.prevDensity !== this.cellDensity()) {
    this.prevDensity = this.cellDensity();
  ...
}
```

jbrundage replied:
Yes, valid point, I'll modify this as well.

GordonSmith commented on src/layout/Grid.js:
```diff
@@ -422,27 +445,35 @@
-                    dropCells.push({ x: colIdx, y: rowIdx });
-                }
+        if(this.designMode()){
+            var c_canvas = document.createElement('canvas');
```
Do you need to create a new canvas every update cycle?

jbrundage replied:
No it doesn't need to... Good point, I'll modify this to only redraw the canvas when necessary.

GordonSmith replied:
Or at least remove it it at the end.

jbrundage commented:
@GordonSmith @jchambers-ln I've added the cell selection feature to Grid's designMode. Just click on a draggable cell while in design mode and the Property Editor will be filtered. When none are selected the PE goes back to normal.

Give it a try:
https://rawgit.com/jbrundage/Visualization/Grid_Optimizing_designMode/demos/dermatology.html?src/layout/Grid&designMode=true

#636

GordonSmith commented:
Resizing (bigger) seems to be broken now.

jbrundage commented:
I see what you mean... re-sizing the window doesn't trigger the canvas redraw.

jbrundage commented:
@GordonSmith I've fixed the re-sizing issue. Please Review

GordonSmith commented:
@jbrundage what about my other comments - I think for the first release you should only support single selection.

jbrundage commented:
Alright, I'll disable multi-select.

jbrundage commented:
@GordonSmith I've removed the multi-select.

I've added a publish param array ''excludeTags" which defaults to ["Advanced","Private"] and moved Surface's "widget" param to the top of the list. #753

GordonSmith commented:
@jbrundage Resizing still not working.

GordonSmith commented:
@jbrundage I would put title above widget and auto collapse the PropertyEditor branch.

jbrundage commented:
@GordonSmith Fixed and done and done.

GordonSmith commented:
@jbrundage Can you rebase on master to fix that lint error?

GordonSmith commented:
Closing in favor of:  https://github.com/hpcc-systems/Visualization/pull/776
